agent/Retriever.py

```diff
@@ -0,0 +1,76 @@
+from llama_index import (
+    VectorStoreIndex,
+    SummaryIndex,
+    SimpleKeywordTableIndex,
+    SimpleDirectoryReader,
+    ServiceContext,
+)
+from llama_index.agent import OpenAIAgent
+from llama_index.schema import IndexNode
+from llama_index.tools import QueryEngineTool, ToolMetadata
+from llama_index.llms import OpenAI
+import os
+from llama_index.retrievers import RecursiveRetriever
+from llama_index.query_engine import RetrieverQueryEngine
+from llama_index.response_synthesizers import get_response_synthesizer
+
+class Retriever:
+    def __init__(self, index_path):
+        
+        self.llm = OpenAI(temperature=0, model="gpt-4-1106-preview")
+    self.service_context = ServiceContext.from_defaults(llm=self.llm)
+    
+    def load_documents(self, files):
+        self.docs = {}
+        for file in files.keys():
+            docs[file] = SimpleDirectoryReader(
+                input_files=[f"./{file}"],
+            ).load_data()
+
+
+
+    def create_agents(self, files, docs, service_context):
+            summary_index = SummaryIndex.from_documents(docs[file], service_context=service_context)
+            
+            # define query engines
+            vector_query_engine = vector_index.as_query_engine()
+            list_query_engine = summary_index.as_query_engine()
+            
+            # define tools
+            query_engine_tools = [
+                QueryEngineTool(
+                    query_engine=vector_query_engine,
+                    metadata=ToolMetadata(name="vector_tool", description=f"Useful for summarization questions related to {file}"),
+                ),
+                QueryEngineTool(
+                    query_engine=list_query_engine,
+                    metadata=ToolMetadata(name="summary_tool", description=f"Useful for retrieving specific context from {file}"),
+                ),
+            ]
+            
+            # build agent
+            function_llm = OpenAI(model="gpt-4-1106-preview")
```

Comment:
Creating a new `OpenAI` instance with the same parameters in `create_agents` method seems redundant when one is already available in `self.llm`. Consider reusing the existing instance if possible.
```suggestion
            function_llm = self.llm
```